feat(totp): add totp rate limits

vbudhram opened the pull request:
Fixes #226 

This PR adds rate limiting to actions that involve checking a TOTP code (only `verifyTotpCode` atm ). Unfortunately, while adding this rate limit, I could not re-use the existing rate limiting structure for ip and email based actions since those limits are defined `config.limits.rateLimitIntervalSeconds`. The TOTP rate limit requirement is 2 checks every 30 seconds (rate the codes are generated).

To support this feature, I decided to build off https://github.com/mozilla/fxa-customs-server/pull/174, which had a lofty goal of refactoring all the things. This PR is a smaller subset that adds support for creating a custom rule for actions and checking limits.

* To maintain backwards comp, a new route is used `/checkWithRule`
* Added a generic record class that stores custom limits and actions for rule
* Records are stored in memcache with key `ip + email + ruleName`

While this doesn't negate the fact we should refactor this service, it does provide the minimum parts needed to support rate limiting TOTP.

@rfk What are your thoughts on this approach?

rfk commented on lib/config/config.js:
```diff
@@ -279,6 +279,37 @@ module.exports = function (fs, path, url, convict) {
       format: 'String',
       env: 'SENTRY_DSN'
+    },
+    customRules: {
```
Naming nit: using "custom" in the sense of "customized" may be confusing when mixed with "customs" in the sense of "customs and border security", I wonder if we can find an equivalent term that's less ambiguous.

rfk commented on lib/config/config.js:
```diff
@@ -279,6 +279,37 @@ module.exports = function (fs, path, url, convict) {
+            env: 'TOTP_CODE_RULE_PERIOD_MS'
+          },
+          rateLimitIntervalMs: {
+            doc: 'how long long rate limit is applied',
```
nit: "long long"

rfk commented on lib/server.js:
```diff
@@ -386,6 +387,91 @@ module.exports = function createServer(config, log) {
+    const customRules = config.customRules
+    const checkRules = Object.keys(customRules).filter((key) => {
+      return customRules[key].actions.indexOf(action) > -1
+    })
```
I suggest we pre-compute this at load time into a hash that maps action names to list of applicable rules, rather than doing a nested loop here on each request.

rfk commented on lib/server.js:
```diff
@@ -386,6 +387,91 @@ module.exports = function createServer(config, log) {
+    // Retrieve the record for each rule and check if `action` has been rate limited
+    const recordResults = []
+    checkRules.forEach((ruleName) => {
+      const recordKey = ip + email + ruleName
```
If it's a relatively small change, let's consider hashing this down to something with a known fixed length, to avoid https://bugzilla.mozilla.org/show_bug.cgi?id=1365486

rfk commented on lib/server.js:
```diff
@@ -386,6 +387,91 @@ module.exports = function createServer(config, log) {
+    }
+
+    // Retrieve the record for each rule and check if `action` has been rate limited
+    const recordResults = []
```
The `P.each` helper might make this iteration cleaner.

vbudhram commented:
Pulling WIP off of this since I think it is ready. It can be tested against https://github.com/mozilla/fxa-auth-server/pull/2386, just make sure you comment out `customsUrl` in `dev.json` before running.

Currently, this rule will only apply when verifying a TOTP code and allows 2 attempts every 30s.

@mozilla/fxa-devs r?

vbudhram commented on lib/server.js:
```diff
@@ -386,6 +406,88 @@ module.exports = function createServer(config, log) {
+    const recordResults = []
+    return P.each(checkRules, (ruleName) => {
+      let retryAfter = null
+      const recordKey = ip + utils.createHashHex(email, ruleName)
```
Is there any value in keeping IP as part of memcache key? For example, would we ever need to delete user defined rules for a specific IP?

rfk replied:
I can't think of a reason why we'd want to do this off the top of my head.  I also don't recall there being a way to delete keys from memcached by prefix, so in order to delete the key, you'd have to know all the inputs to the key calculation anyway.

I think we're more likely to want to have the ruleName in cleartext than either the IP or email, which would make this something like `recordKey = ruleName + ":" + utils.createHashHex(ip, email)` or similar.

rfk commented on lib/server.js:
```diff
@@ -386,6 +406,88 @@ module.exports = function createServer(config, log) {
     }
   )
 
+  api.post('/checkCustomRule', function(req, res, next) {
```
nit: this api still says "custom rule", which no longer matches the terminology used in the code here.

rfk replied:
By moving this into its own route, we bypass all the checks in the usual `/check` route.  IIUC that's mostly what we want, but I wonder if we've lost any that we should keep, such a the ip-reputation-service checking, the ip-blocklist checking, or the explicit "block this IP address" API call offered by the `/blockIp` route?

rfk commented on lib/server.js:
```diff
@@ -386,6 +406,88 @@ module.exports = function createServer(config, log) {
+        })
+        .then(() => {
+          recordResults.push({
+            block: retryAfter > 0,
```
Do you actually use this `block` key in the logic below?  I wonder if this would be simpler as just a list of `retryAfter` integers, rather than a list of rich objects with a `retryAfter` key.

rfk commented on lib/utils.js:
```diff
@@ -0,0 +1,18 @@
+    const hash = crypto.createHash('sha256')
+    const args = [...arguments]
+    args.forEach((arg) => {
+      hash.update(arg)
```
Please include a separator between each argument here, preferably one that won't appear in any of the input strings, such as `'\00'`.  Otherwise, the hash for e.g. `{ ip: "1.2.3.45", email: "test@example.com }` could wind up being the same as the hash for `{ ip: "1.2.3.4", email: "5test@example.com" }`, which would cause us to incorrectly treat those as the same.

rfk commented on test/remote/too_many_totp_codes.js:
```diff
@@ -0,0 +1,96 @@
+      t.fail(err)
+      t.end()
+    })
+})
```
I think it'd be worth adding another test here, to ensure that the rate-limiting is independent of other rate-limiting actions.  For example, that a user can still attempt `accountLogin` actions after be rate-limited for totp verification, and conversely that a user rate-limited for `accountLogin` actions can still do totp verification attempts.

vbudhram commented:
@rfk After thinking it over, it seems reasonable to put this logic into the `/check` route. I updated the code to do that and refactored the route a little bit.

Mind another review at your convenience? Probably without [whitespace](https://github.com/mozilla/fxa-customs-server/pull/233/files?w=1).

rfk commented on lib/utils.js:
```diff
@@ -0,0 +1,18 @@
+    const hash = crypto.createHash('sha256')
+    const args = [...arguments]
+    args.forEach((arg) => {
+      hash.update(arg + '\\00')
```
I think this should be `'\0'` in order to get a literal null byte, rather than the string "\00".
